### PR TITLE
[refactor] pipe: simplify balance and module checks

### PR DESCRIPTION
**Pull request by msbaines**
Also, remove run-time type checks. mypy is already checking types.
However, a user could still pass the wrong types and not use
type-checking.

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/nn/pipe/multiprocess_pipe.py [hunk 1]
```diff
@@ -53,85 +53,22 @@
...
-
-    if filter_unique:
-        module_len = len(set(map(id, module)))
-    else:
-        module_len = len(module)
-
-    if module_len != sum(balance):
-        raise BalanceError(
+def check_balance(module: Union[nn.Sequential, List[LazyModule]], balance: List[int]) -> None:
+    if len(module) != sum(balance):
```
Comment (sidgoyal78) [inline fairscale/nn/pipe/multiprocess_pipe.py:62]: I'm curious why this isn't `len(set(map(id, module))) != len(module)`, similar to the condition in `verify_module()`?

**Review by sidgoyal78 — COMMENTED** [review]
Thanks for the refactor. I have put an inline question.

**Review by anj-s — APPROVED** [review]

File: fairscale/nn/pipe/multiprocess_pipe.py (hunk 1 above)
Comment (msbaines) [reply fairscale/nn/pipe/multiprocess_pipe.py:62]: One verifies that there are no duplicates in the list of layers. While the other verifies that the size of the list of layers matches the size of balance.

Comment (sidgoyal78) [reply fairscale/nn/pipe/multiprocess_pipe.py:62]: Yeah, but from the code that you removed (at one place), the condition that I mentioned was getting called?

**Pull request merged**